compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp

```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
       return failure();
     }
 
-    // k == 1
-    if (lhsType.getDimSize(dimensionNumbers.getLhsContractingDimensions()[0]) !=
-            1 ||
-        rhsType.getDimSize(dimensionNumbers.getRhsContractingDimensions()[0]) !=
-            1) {
-      return failure();
+    auto is_single_k = [&]() {
+      return lhsType.getDimSize(
+                 dimensionNumbers.getLhsContractingDimensions()[0]) == 1 &&
+             rhsType.getDimSize(
+                 dimensionNumbers.getRhsContractingDimensions()[0]) == 1;
+    };
+
+    auto is_single_m_n = [&]() {
+      return lhsType.getDimSize(lhsType.getRank() - 2) == 1 ||
+             rhsType.getDimSize(rhsType.getRank() - 1) == 1;
+    };
+
+    auto resultRank = op.getType().getRank();
+    auto resultShape = op.getType().getShape();
+    auto resultElementTy = op.getType().getElementType();
+    auto has_batch = dimensionNumbers.getLhsBatchingDimensions().size() != 0 ||
+                     dimensionNumbers.getRhsBatchingDimensions().size();
+    auto B = has_batch ? resultShape[0] : 0;
+    auto M = lhsType.getDimSize(lhsType.getRank() - 2);
+    auto N = rhsType.getDimSize(rhsType.getRank() - 1);
+    auto LK =
+        lhsType.getDimSize(dimensionNumbers.getLhsContractingDimensions()[0]);
+    auto RK =
+        rhsType.getDimSize(dimensionNumbers.getRhsContractingDimensions()[0]);
+    auto K = std::max(LK, RK);
+
+    if (is_single_k()) {
```

Comment:
It's not necessary to check k == 1. Just lowering it to mhlo.reduce with reduce rank size == 1, the mhlo.reduce will be simplified to reshape by SimplifyReduceToReshape pattern.